_posts/2025-04-28-sparse-autodiff.md

```diff
@@ -502,24 +514,49 @@ in this case addition, multiplication and the sign function.
 </div>
 
 As discussed in the previous section,
-all inputs are seeded with their respective input index sets.
-Figure 14 annotates these index sets on the edges of the computational graph.
-Our system for sparsity detection must now perform an **alternative evaluation of our computational graph**.
-Instead of computing the original function, 
-each operator must correctly propagate and accumulate the index sets of its inputs, 
-depending on whether an operator has a non-zero derivative or not.  
-
-Since addition and multiplication globally have non-zero derivatives with respect to both of their inputs, 
+all inputs are initialized with their respective singleton index sets.
+The sparsity detection system must walk through the computational graph in topological order.
+Each individual operator in the graph is "overloaded" to behave in a different way during sparsity detection.
+Instead of computing the output value, 
+it must correctly propagate and accumulate the index sets of its inputs, 
+depending on whether it has a non-zero derivative or not.  
+
+Figure 14 annotates index sets on the edges of the computational graph.
+Since addition and multiplication have non-zero derivatives with respect to both of their inputs, 
 the index sets of their inputs are accumulated and propagated. 
 The sign function has a zero-valued derivatives for any input value. 
 It therefore doesn't propagate the index set of its input. 
 Instead, it returns an empty set.
 
-*TODO: switch to multivariate function, quickly discuss resulting Jacobian.*
+<!-- *TODO: switch to multivariate function, quickly discuss resulting Jacobian.* -->
 <!-- TODO -->
 
-### Matrix coloring
+<!-- Coloring techniques? -->
+## Coloring
+
+Once we have detected a sparsity pattern, our next goal is to figure out how to group the columns (or rows) of the Jacobian.
+The columns (or rows) in each group will be evaluated simultaneously with a single JVP (or VJP).
+If they are mutually orthogonal, then this gives all the necessary information to retrieve every nonzero coefficient of the matrix.
+
+### Graph formulation
+
+Luckily, this can be reformulated as a graph coloring problem, which is very well studied.
+Let us build a graph $\mathcal{G} = (\mathcal{V}, \mathcal{E})$ such that each column is a vertex of the graph, and two vertices are connected iff their respective columns share a non-zero index.
```

Comment:
`\mathcal{V}` and `\mathcal{E}` should be introduced.

Reply:
```suggestion
Let us build a graph $\mathcal{G} = (\mathcal{V}, \mathcal{E})$ with vertex set $\mathcal{V}$ and edge set $\mathcal{E}$, such that each column is a vertex of the graph, and two vertices are connected iff their respective columns share a non-zero index.
```